subprojects/core/src/main/java/org/gradle/api/internal/tasks/DefaultTaskOutputs.java

```diff
@@ -103,18 +130,28 @@ public boolean isCacheAllowed() {
 
     @Override
     public void cacheIf(final Spec<? super Task> spec) {
+        cacheIf("Task output is not cacheable", spec);
+    }
+
+    @Override
+    public void cacheIf(final String message, final Spec<? super Task> spec) {
         taskMutator.mutate("TaskOutputs.cacheIf(Spec)", new Runnable() {
             public void run() {
-                cacheIfSpec = cacheIfSpec.and(spec);
+                cacheIfSpecs.add(new SpecWithDescription<TaskInternal>(spec, message));
```

Comment:
Maybe add something...

`Spec Specs.describe(String message, Spec spec)`

So we could do something like...

`cacheIfSpec = cacheIfSpec.and(Specs.describe(message, spec))`

And we could make the `CompositeSpec` build a description for the whole chain?